internal/scan/scan.go

```diff
@@ -0,0 +1,181 @@
+		slog.Debug("no events found for block range", "from_block", fromBlock.Int64(), "to_block", toBlock.Int64())
+		return nil, nil
+	}
+
+	contractAbi, err := abi.JSON(strings.NewReader(ERC721.ERC721MetaData.ABI))
+	if err != nil {
+		return nil, fmt.Errorf("error instantiating ABI: %w", err)
+	}
+	var parsedEvents []Event
+	for i := range eventLogs {
+		slog.Info("scanning event", "block", eventLogs[i].BlockNumber, "txHash", eventLogs[i].TxHash)
+		switch eventLogs[i].Topics[0].Hex() {
+		case eventTransferSigHash:
+			transfer, err := parseTransfer(&eventLogs[i], &contractAbi)
+			if err != nil {
+				return nil, err
+			}
+			parsedEvents = append(parsedEvents, transfer)
+			slog.Info("received event", eventTransferName, transfer)
+		case eventApprovalSigHash:
+			approval, err := parseApproval(&eventLogs[i], &contractAbi)
+			if err != nil {
+				return nil, err
+			}
+			parsedEvents = append(parsedEvents, approval)
+			slog.Info("received event", eventApprovalName, approval)
+		case eventApprovalForAllSigHash:
+			approvalForAll, err := parseApprovalForAll(&eventLogs[i], &contractAbi)
+			if err != nil {
+				return nil, err
+			}
+			parsedEvents = append(parsedEvents, approvalForAll)
+			slog.Info("received event", eventApprovalForAllName, approvalForAll)
+		default:
+			slog.Warn("unrecognized event", "eventType", eventLogs[i].Topics[0].String())
+		}
+	}
+
+	return parsedEvents, nil
+}
+
+func (s scanner) filterEventLogs(ctx context.Context, firstBlock, lastBlock *big.Int) ([]types.Log, error) {
+	return s.cli.FilterLogs(ctx, ethereum.FilterQuery{
+		FromBlock: firstBlock,
+		ToBlock:   lastBlock,
+		Addresses: []common.Address{s.contract},
+	})
+}
+
+func parseTransfer(eL *types.Log, contractAbi *abi.ABI) (EventTransfer, error) {
+	var transfer EventTransfer
+	err := unpackIntoInterface(&transfer, contractAbi, eL)
+	if err != nil {
+		return transfer, err
+	}
+	transfer.From = common.HexToAddress(eL.Topics[1].Hex())
+	transfer.To = common.HexToAddress(eL.Topics[2].Hex())
+	transfer.TokenId = eL.Topics[3].Big()
+	return transfer, nil
+}
+
+func parseApproval(eL *types.Log, contractAbi *abi.ABI) (EventApproval, error) {
+	var approval EventApproval
+	err := unpackIntoInterface(&approval, contractAbi, eL)
+	if err != nil {
+		return approval, err
+	}
+	approval.Owner = common.HexToAddress(eL.Topics[1].Hex())
+	approval.Approved = common.HexToAddress(eL.Topics[2].Hex())
+	approval.TokenId = eL.Topics[3].Big()
+	return approval, nil
+}
+
+func parseApprovalForAll(eL *types.Log, contractAbi *abi.ABI) (EventApprovalForAll, error) {
+	var approvalForAll EventApprovalForAll
+	err := unpackIntoInterface(&approvalForAll, contractAbi, eL)
+	if err != nil {
+		return approvalForAll, err
+	}
+	approvalForAll.Owner = common.HexToAddress(eL.Topics[1].Hex())
+	approvalForAll.Operator = common.HexToAddress(eL.Topics[2].Hex())
+	return approvalForAll, nil
+}
+
+func unpackIntoInterface(e Event, contractAbi *abi.ABI, eL *types.Log) error {
```

Comment:
I think it would be better pass eventTransferName as an parameter to the function.
It would be clearer and more robust.